Add quarter helices to various tracks

mixiate opened the pull request:
This adds unbanked and banked quarter helices to the Alpine, Corkscrew, Giga, Looping, Mine Ride, Mini, Single Rail, Stand Up and Twister track types.

<img width="932" height="603" alt="quarterhelices" src="https://github.com/user-attachments/assets/cf5ed8d4-fb3e-46bf-9282-f0af19ab35d7" />

The clearances on these track pieces are very low (they are arguably slightly too low on the inverted pieces too), let me know how much of an issue you think this is.

This adds new strings for the banked quarter helices only, to differentiate them from the half helices in the special track piece menu.

Gymnasiast commented:
@ZeeMaji You know the drill ;-)

brisingr25 commented:
These are so, so great! I've been testing some more, I've found a couple things:

Tunnel clearances are too low for both banked and unbanked versions:

![tunnel2](https://github.com/user-attachments/assets/4eb74f82-587b-40c1-9bdb-f362522044f1)

Also maybe this support gap could be improved:

![tunnel1](https://github.com/user-attachments/assets/4066216b-2425-4b3e-aa7d-370014f42a0a)

Lighting could be improved on the twister track here:

![light1](https://github.com/user-attachments/assets/44b02258-593c-4b12-8756-87bea43e034a)

And then there's this sort of connection, but as we already discussed the vanilla track is full of these things so I don't know if it's worth fixing:

![cats3](https://github.com/user-attachments/assets/538acf35-6519-48f4-af09-87aa58bc4320)

mixiate commented:
Can't change the clearances without being a backwards incompatible change. I think you could do the land slope as that would just be more permissive.
Couldn't do tile height either without breaking many things. Could easily live without it if it were possible to change the clearances on the existing piece.

brisingr25 commented:
Did some more testing, all seems fine to me! I can only find some minor things if I get crazy combining large helixes, quarter helixes and level turns, but it's all about the clearances again. In the same style, this support glitch, but it probably isn't related with this PR:

![support1](https://github.com/user-attachments/assets/3668e2bd-6d0a-4023-b3ef-639eb538c6f6)

What I find somewhat confusing is the naming of the pieces on the special pieces list:

![list](https://github.com/user-attachments/assets/bf90cdad-a909-49e5-8d7a-4ad37284c595)

Because they take the same name from the vanilla inverted and suspended coasters, when you have both banked and unbanked quarter helixes available on the same coaster the names get confusing when building. Why does the unbanked quarter helix not contain "quarter" or "large" in its name? Should we include the "large" specification on quarter helixes since there's no small size? It also gets confusing with the normal non-quarter helixes. Maybe it would be a good idea to propose a renaming of all helixes, to something like this for example:

Helix up (small)
Helix down (small)
Helix up (large)
Helix down (large)
Quarter helix up (banked)
Quarter helix down (banked)
Quarter helix up (left)
Quarter helix up (right)
Quarter helix down (left)
Quarter helix down (right)

But then there's the spiral coaster. If at some point a piece like the spiral lift hill but without the lift is added to other coasters, it'll get even more confusing... Then maybe we should differenciate between large and small quarter helixes, I suppose.

mixiate commented:
I've added the hybrid coaster. I'm not going to add another track type to this PR.
They needed a different paint implementation. I didn't want to copy paste modify another load of code so I decided to implement it, as much as is possible with how things are now, in the way I have in mind for refactoring all track paint functions and moving towards track objects. More details in the main post.

brisingr25 commented:
All looks good to me, except these floating pixels on the hybrid that weren't there before:

![hybrid](https://github.com/user-attachments/assets/7515b5e4-08fe-4c7c-9a1c-e110e14127d2)

mixiate commented:
This might be a horrible idea but just throwing things out there.
Add compatibility clearance to track sequence data that is used when placing track designs.
https://github.com/OpenRCT2/OpenRCT2/compare/develop...mixiate:OpenRCT2:compatibility-clearance-test

xrgtmx commented:
Any plans on including the Dinghy Slide track to this PR or the future?

mixiate commented:
I've rebased this, and added the changes for the track piece names and the clearance heights.

This changes the names of half helices and quarter helices. These are the names I went with, but I realise they may not be the best so let me know your thoughts.
<img width="313" height="333" alt="unbankedquarterhelixname" src="https://github.com/user-attachments/assets/1bf38df9-f90f-4a55-8b29-0a5c52e0427b" /><img width="380" height="317" alt="bankedhalfandquarterhelixnames" src="https://github.com/user-attachments/assets/648b152a-24b1-4dda-8c30-367b8e5ccbb9" />

To recap the situation with the clearances, they're too low even on the existing inverted pieces, allowing you to build things like this without cheats:
<img width="323" height="238" alt="Untitled" src="https://github.com/user-attachments/assets/9f5c4499-2a71-4734-8d49-2165b1966ab0" /><img width="532" height="435" alt="Untitled" src="https://github.com/user-attachments/assets/d4a74bfe-4589-4051-a004-8408158ad0db" />
This raises all tiles so that they're 1 unit of height higher than they were.
This also changes the slope underneath the last tile so that you can build them over land like this:
<img width="478" height="434" alt="slopeunderneath" src="https://github.com/user-attachments/assets/9911af26-3eb4-4e2a-b730-22010fbf84ec" />

brisingr25 commented:
All looks good to me! Maybe lose the "(medium)" on the banked quarter helixes? I think it works as well without it.

mixiate commented:
I'm not opposed to removing the medium. I thought of future proofing it if small versions are ever added. However I don't actually plan on making those pieces as the angle from flat is too great.
Also the unbanked ones don't have the size in them either so it would be more consistent.

brisingr25 commented:
Well, we do have a precedent of what a small unbanked quarter helix looks like, the spiral lift hill:

<img width="1664" height="1664" alt="quarter" src="https://github.com/user-attachments/assets/73c3b8a3-3892-4976-9fbc-46d78d6f2df4" />

So maybe it wouldn't be so weird to add a small unbanked quarter helix piece in the future... I don't know if a banked one would work, though.

Then again, half helices being named small and large would have to be changed if large curve radius half helixes or quarter helices are ever added, and the current large half helices would then be medium half helices, and it would coincide with the medium quarter helix naming you proposed. But I'm overcomplicating things, so maybe just drop the "medium" for now!

brisingr25 commented:
Everything looks good to me! I think the pieces are ready for merging, very nice work Mix!

Gymnasiast commented on src/openrct2/ride/TrackPaint.h:
```diff
@@ -551,6 +553,93 @@ void TrackPaintUtilDiagTilesPaintExtra(
+static constexpr uint64_t kLeftQuarterHelixLargeUpSpriteMap = 0b01010001010001010100010100010101000101000101010001010001;
+static constexpr uint64_t kRightQuarterHelixLargeUpSpriteMap = 0b01010001010001010100010100010101000101000101010001010001;
+static constexpr uint64_t kLeftQuarterBankedHelixLargeUpSpriteMap = 0b01010001010001111100010100011010001010001001010001010011;
+static constexpr uint64_t kRightQuarterBankedHelixLargeUpSpriteMap = 0b01010001010011101000101000101111000101000101010001010001;
```
These are really obfuscated. They seem like they are a bitset, but I can’t make heads or tails of it. It basically looks like the kind of structure we’ve been trying very hard to move away from.

mixiate replied:
Yeah it's bits, 1 means that there is a sprite for that sequence/view/back or front.
It's data though. It's supposed to come with the track type, and different tracks can use different types (only hybrid in this PR). Maybe it's possible to write a const function to produce it, but it's not something that's supposed to be hardcoded.

I can't say whether it's the best way to go about things, it has some drawbacks, but I wanted to try it. When everything is hardcoded it's not possible to try different things en masse, all you can do is assume.
The most likely alternatives would be to stick with the hardcoded switch statements or equivalent, or use [arrays like this](https://github.com/mixiate/OpenRCT2/blob/b0166f764f7109f7366be3a476657417f9486b74/src/openrct2/paint/track/coaster/LatticeTriangle.h#L354-L392).

Gymnasiast replied:
I don’t understand. Surely this could be written in a more readable format? I'm certainly not advocating putting large switch statements in, I just want these constants to be readable.

mixiate replied:
Yeah it can be, it's not really supposed to be readable and like I said the reason why I didn't calculate it at compile time is because eventually it would be not hardcoded. But it can be done. How's this:

```
const size_t kSpriteMapSpriteCount = 2;

template<bool flipXAxis, size_t sequenceCount>
constexpr uint64_t createSpriteMap(
    const std::array<std::array<std::array<bool, kSpriteMapSpriteCount>, sequenceCount>, kNumOrthogonalDirections>& array)
{
    const size_t directionOffset = sequenceCount * kSpriteMapSpriteCount;
    uint64_t spriteMap = 0;
    for (size_t direction = 0; direction < kNumOrthogonalDirections; direction++)
    {
        for (size_t sequence = 0; sequence < sequenceCount; sequence++)
        {
            for (size_t sprite = 0; sprite < kSpriteMapSpriteCount; sprite++)
            {
                if (array[direction][sequence][sprite] == true)
                {
                    size_t directionIndex = direction;
                    if constexpr (flipXAxis)
                    {
                        directionIndex = kNumOrthogonalDirections - 1 - direction;
                    }
                    const uint64_t shift = sprite + (sequence * kSpriteMapSpriteCount) + (directionIndex * directionOffset);
                    const uint64_t bit = 0b1LL << shift;
                    spriteMap |= bit;
                }
            }
        }
    }
    return spriteMap;
}

static constexpr std::array<std::array<std::array<bool, kSpriteMapSpriteCount>, 7>, kNumOrthogonalDirections>
    kLeftQuarterHelixLargeUpSpriteMapArray = { { { { { { true, false } },
                                                     { { false, false } },
                                                     { { true, false } },
                                                     { { true, false } },
                                                     { { false, false } },
                                                     { { true, false } },
                                                     { { true, false } } } },
                                                 { { { { true, false } },
                                                     { { false, false } },
                                                     { { true, false } },
                                                     { { true, false } },
                                                     { { false, false } },
                                                     { { true, false } },
                                                     { { true, false } } } },
                                                 { { { { true, false } },
                                                     { { false, false } },
                                                     { { true, false } },
                                                     { { true, false } },
                                                     { { false, false } },
                                                     { { true, false } },
                                                     { { true, false } } } },
                                                 { { { { true, false } },
                                                     { { false, false } },
                                                     { { true, false } },
                                                     { { true, false } },
                                                     { { false, false } },
                                                     { { true, false } },
                                                     { { true, false } } } } } };

static constexpr std::array<std::array<std::array<bool, kSpriteMapSpriteCount>, 7>, kNumOrthogonalDirections>
    kLeftQuarterBankedHelixLargeUpSpriteMapArray = { { { { { { true, true } },
                                                           { { false, false } },
                                                           { { true, false } },
                                                           { { true, false } },
                                                           { { false, false } },
                                                           { { true, false } },
                                                           { { true, false } } } },
                                                       { { { { false, true } },
                                                           { { false, false } },
                                                           { { false, true } },
                                                           { { false, true } },
                                                           { { false, false } },
                                                           { { false, true } },
                                                           { { false, true } } } },
                                                       { { { { true, false } },
                                                           { { false, false } },
                                                           { { true, false } },
                                                           { { true, false } },
                                                           { { false, false } },
                                                           { { true, true } },
                                                           { { true, true } } } },
                                                       { { { { true, false } },
                                                           { { false, false } },
                                                           { { true, false } },
                                                           { { true, false } },
                                                           { { false, false } },
                                                           { { true, false } },
                                                           { { true, false } } } } } };

static constexpr uint64_t kLeftQuarterHelixLargeUpSpriteMap = createSpriteMap<false>(kLeftQuarterHelixLargeUpSpriteMapArray);
static constexpr uint64_t kRightQuarterHelixLargeUpSpriteMap = createSpriteMap<true>(kLeftQuarterHelixLargeUpSpriteMapArray);
static constexpr uint64_t kLeftQuarterBankedHelixLargeUpSpriteMap = createSpriteMap<false>(
    kLeftQuarterBankedHelixLargeUpSpriteMapArray);
static constexpr uint64_t kRightQuarterBankedHelixLargeUpSpriteMap = createSpriteMap<true>(
    kLeftQuarterBankedHelixLargeUpSpriteMapArray);
```

and for the hybrid:

```
static constexpr std::array<std::array<std::array<bool, kSpriteMapSpriteCount>, 7>, kNumOrthogonalDirections>
    kLeftQuarterHelixLargeUpSpriteMapArray = { { { { { { true, false } },
                                                     { { false, false } },
                                                     { { true, false } },
                                                     { { true, false } },
                                                     { { false, false } },
                                                     { { true, false } },
                                                     { { true, false } } } },
                                                 { { { { true, false } },
                                                     { { false, false } },
                                                     { { true, false } },
                                                     { { true, false } },
                                                     { { false, false } },
                                                     { { true, false } },
                                                     { { true, false } } } },
                                                 { { { { true, false } },
                                                     { { false, false } },
                                                     { { true, false } },
                                                     { { true, false } },
                                                     { { false, false } },
                                                     { { true, false } },
                                                     { { true, false } } } },
                                                 { { { { true, false } },
                                                     { { true, false } },
                                                     { { true, false } },
                                                     { { true, false } },
                                                     { { false, false } },
                                                     { { true, false } },
                                                     { { true, false } } } } } };

static constexpr std::array<std::array<std::array<bool, kSpriteMapSpriteCount>, 7>, kNumOrthogonalDirections>
    kLeftQuarterBankedHelixLargeUpSpriteMapArray = { { { { { { true, true } },
                                                           { { false, false } },
                                                           { { true, false } },
                                                           { { true, false } },
                                                           { { false, false } },
                                                           { { true, false } },
                                                           { { true, false } } } },
                                                       { { { { true, true } },
                                                           { { false, false } },
                                                           { { true, true } },
                                                           { { true, true } },
                                                           { { false, false } },
                                                           { { true, true } },
                                                           { { true, true } } } },
                                                       { { { { true, false } },
                                                           { { false, false } },
                                                           { { true, false } },
                                                           { { true, false } },
                                                           { { false, false } },
                                                           { { true, true } },
                                                           { { true, true } } } },
                                                       { { { { true, false } },
                                                           { { true, false } },
                                                           { { true, false } },
                                                           { { true, false } },
                                                           { { false, false } },
                                                           { { true, false } },
                                                           { { true, false } } } } } };

static constexpr uint64_t kLeftQuarterHelixLargeUpSpriteMap = createSpriteMap<false>(
    kLeftQuarterHelixLargeUpSpriteMapArray);
static constexpr uint64_t kRightQuarterHelixLargeUpSpriteMap = createSpriteMap<true>(
    kLeftQuarterHelixLargeUpSpriteMapArray);
static constexpr uint64_t kLeftQuarterBankedHelixLargeUpSpriteMap = createSpriteMap<false>(
    kLeftQuarterBankedHelixLargeUpSpriteMapArray);
static constexpr uint64_t kRightQuarterBankedHelixLargeUpSpriteMap = createSpriteMap<true>(
    kLeftQuarterBankedHelixLargeUpSpriteMapArray);
```

Gymnasiast replied:
I’m afraid I _still_ don’t understand what information is supposed to be contained in there. The nested arrays and naked number 7 do not really help.

mixiate replied:
This track piece has 7 tiles, I'd use the track element data sequence count but it's not constexpr.

It's per view rotation, per track sequence, then true/false for whether there's a sprite behind the train and in front of the train.

Gymnasiast replied:
Ah, that makes it a bit clearer. I'll see if I can whip something up myself.

Gymnasiast replied:
Right, could you:

1. create a local constant for the number of a segments in a quarter helix (e.g. `static constexpr uint8_t kNumQuarterHelixSegments = 7;`)
2. Use a C-style array (which may sound odd, for multidimensional arrays, it’s easier to read. E.g. `kLeftQuarterBankedHelixLargeUpSpriteMapArray[kNumOrthogonalDirections][kNumQuarterHelixSegments][kSpriteMapSpriteCount] puts all the dimensions easily together).

I think that is a good compromise, which should also be easy to implement for you.

Gymnasiast commented on src/openrct2/park/Legacy.cpp:
```diff
@@ -2982,6 +2982,35 @@ bool TrackTypeMustBeMadeInvisible(ride_type_t rideType, TrackElemType trackType,
+         || rideType == RIDE_TYPE_SPIRAL_ROLLER_COASTER || rideType == RIDE_TYPE_MINE_RIDE
+         || rideType == RIDE_TYPE_HYBRID_COASTER || rideType == RIDE_TYPE_STEEPLECHASE
+         || rideType == RIDE_TYPE_MULTI_DIMENSION_ROLLER_COASTER)
+        && parkFileVersion < kParkFileVersionUprightQuarterHelices)
```
Put the `parkFileVersion` check first, it avoids lots of useless checks in for newer saves.

Gymnasiast commented on data/language/en-GB.txt:
```diff
@@ -3851,3 +3851,5 @@ STR_7024    :Makes the whole ride invisible
 STR_7026    :Makes the whole ride visible
 STR_7027    :{MOVE_X}{29}{STRINGID}
+STR_7028    :Quarter Helix up
+STR_7029    :Quarter Helix down
```
Are these new strings necessary? It seems to duplicate the STR_1679 - STR_1682 block, but without the left/right postfixes.

Gymnasiast replied:
On second though, I can see why, I think. Makes the Special dropdown a bit less cluttered.

Gymnasiast reviewed (commented):
Looks good so far, I do have two line notes.

Gymnasiast commented:
@brisingr25 Could you test if nothing has regressed, as well as test the newly added Steeplechase quarter helix?

brisingr25 commented:
I've tested every track type and it all seems fine to me! The new multi-dim pieces are fine, and the steeplechase are as well, the only thing that I've found is this connection that could maybe be improved, but I'm aware that this track type has many instances like this in vanilla, or even worse, so it can be ignored:

![steep1](https://github.com/user-attachments/assets/2bd505ec-dce9-4990-9dd5-0c33f068f829)

Gymnasiast commented:
If I look at the 25-up-to-flat piece, it seems that it is also misaligned with the 25-up pieces before.

What do you think @mixiate ?

brisingr25 commented:
With the latest fix, everything is fine to me and looks ready for merging! Great work once again, Mix!

mixiate commented:
That connection can be fixed without breaking anything else.

<img width="408" height="324" alt="Untitled" src="https://github.com/user-attachments/assets/b2cd58a2-1d81-4219-ad45-4b3397cba785" />

Looks like it's just the lighting on the slopes making them look a little misaligned. As usual, the lighting in general is all a bit weird across different pieces. It's hard to tell what's going on sometimes, and always hard to match anything. I've tried to improve it a little bit too.

Thanks for testing again.

Will need to rebase for the network version.

Gymnasiast commented:
Excellent. Please do that and I’ll merge it.

mixiate commented:
Rebased and folded the updated steeplechase sprites in to the main commit in case you want to merge without squashing.
